File: Exercism/java/sieve/src/main/java/Sieve.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Created by cajetan on 5/17/17.
 */

public class Sieve {
    private int upTo;

    public Sieve(int upTo) {
        this.upTo = upTo;
    }

    List<Integer> getPrimes() {
        List<Integer> result = new ArrayList<>();

        if (upTo == 2) {
            result.add(2);
            return result;
        }

        boolean[] prime_mask = new boolean[upTo];
        Arrays.fill(prime_mask, Boolean.TRUE);

        int i = 2;

        while (Math.pow(i, 2) < upTo) {
            if (prime_mask[i]) {
                int j = i*2;
                while (j < upTo) {
                    prime_mask[j] = false;
                    j += i;
                }
            }

            i++;
        }

        prime_mask[0] = false;
        prime_mask[1] = false;

        for (i = 0 ; i < prime_mask.length ; i++) {
            if (prime_mask[i])
                result.add(i);
        }

        return result;
    }
}

File: Exercism/c/sieve/src/sieve.c
#include <stdio.h>
#include <stdbool.h>
#include <math.h>
#include "sieve.h"

unsigned int sieve(const unsigned int limit, primesArray_t primes) {
    if (limit <= 1)
        return 0;

    int prime_count = 0;

    bool prime_mask[limit+1];
    for (unsigned i = 0 ; i < limit+1 ; ++i)
        prime_mask[i] = true;

    unsigned i = 2;
    while (pow(i, 2) < limit) {
        if (prime_mask[i]) {
            unsigned j = i*2;
            while (j <= limit) {
                prime_mask[j] = false;
                j += i;
            }
        }
        i++;
    }

    prime_mask[0] = false;
    prime_mask[1] = false;

    for (unsigned i = 0 ; i < limit+1 ; ++i) {
        if (prime_mask[i])
            primes[prime_count++] = i;
    }

    return prime_count;
}
